### PR TITLE
ci: Add GH Action to detect compatibility breaking chages

### PR DESCRIPTION
**Pull request by aahung**
### Description of how you validated changes

I created another PR draft to test it: https://github.com/aws/serverless-application-model/pull/2878#issuecomment-1423485136

image

The test GH action run: https://github.com/aws/serverless-application-model/actions/runs/4130070494

### Checklist

- [ ] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [ ] Add/update [transform tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions)
    - [ ] Using correct values
    - [ ] Using wrong values
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)

### Examples?

Please reach out in the comments if you want to add an example. Examples will be 
added to `sam init` through [aws/aws-sam-cli-app-templates](https://github.com/aws/aws-sam-cli-app-templates).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/workflows/check_interface_compatibility.yml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      - name: Backup bin/public_interface.py
+        run: |
+          # Keep a copy of bin/public_interface.py to use on the base branch
+          # So we are using the same bin/public_interface.py to process old/new codebase.
+          cp bin/public_interface.py ${{ runner.temp }}/public_interface.py
+
+      - name: Checkout the base branch
+        uses: actions/checkout@v3
+        with:
+          ref: ${{ github.base_ref }}
```
Comment (hoffa) [inline .github/workflows/check_interface_compatibility.yml:33]: How about checking out to different directories instead? Currently it's kind of confusing with the `cp`s and such.

File: .github/workflows/check_interface_compatibility.yml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+      - name: Extract original public interfaces
+        run: |
+          echo "Original HEAD: $(git rev-parse --short HEAD)"
+          make init
+          # Recover bin/public_interface.py
+          cp ${{ runner.temp }}/public_interface.py bin/public_interface.py
+          bin/public_interface.py extract >> ${{ runner.temp }}/interfaces.original.json
```
Comment (hoffa) [inline .github/workflows/check_interface_compatibility.yml:41]: Why append?

File: .github/workflows/check_interface_compatibility.yml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        id: detect
+        run: |
+          output=$(bash -c '(bin/public_interface.py check ${{ runner.temp }}/interfaces.original.json ${{ runner.temp }}/interfaces.new.json); exit $?')
+          exit_code=$?
+          echo "output=$output"
+          echo "exit_code=$exit_code"
+          echo "output=$output" >> $GITHUB_OUTPUT
+          echo "exit_code=$exit_code" >> $GITHUB_OUTPUT
+
+      - name: Post GitHub PR comment
```
Comment (hoffa) [inline .github/workflows/check_interface_compatibility.yml:60]: Why can't we just let it fail? We use auto-merge so much this won't work unless it's enforced.

File: .github/workflows/check_interface_compatibility.yml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        uses: actions/upload-artifact@v3
+        with:
+          name: interface-json-files
+          path: ${{ runner.temp }}/interfaces.*.json
+          retention-days: 7  # Only storing it for a week in case of debugging needs.
+    
+      - name: Detect compatibility breaking changes
+        id: detect
+        run: |
+          output=$(bash -c '(bin/public_interface.py check ${{ runner.temp }}/interfaces.original.json ${{ runner.temp }}/interfaces.new.json); exit $?')
```
Comment (hoffa) [inline .github/workflows/check_interface_compatibility.yml:53]: Why `bash`? The script should already return the relevant exit code.

File: .github/workflows/check_interface_compatibility.yml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+      - name: Extract original public interfaces
+        run: |
+          echo "Original HEAD: $(git rev-parse --short HEAD)"
+          make init
+          # Recover bin/public_interface.py
+          cp ${{ runner.temp }}/public_interface.py bin/public_interface.py
+          bin/public_interface.py extract >> ${{ runner.temp }}/interfaces.original.json
+
+      - name: Archive interface files for debugging
```
Comment (hoffa) [inline .github/workflows/check_interface_compatibility.yml:43]: Why do we need this? Isn't the script output good enough?

Comment (aahung) [reply .github/workflows/check_interface_compatibility.yml]: This is still draft and you are reviewing the stuff used for my debugging.

File: .github/workflows/check_interface_compatibility.yml (hunk 4 above)
Comment (aahung) [reply .github/workflows/check_interface_compatibility.yml]: not in final version

File: .github/workflows/check_interface_compatibility.yml (hunk 3 above)
Comment (aahung) [reply .github/workflows/check_interface_compatibility.yml:60]: discussed offline, since it is not enforced a comment catches reviewers attention than a small cross, and it is more readable

File: .github/workflows/check_interface_compatibility.yml (hunk 1 above)
Comment (aahung) [reply .github/workflows/check_interface_compatibility.yml]: This approach saves time in `make init`

Comment (aahung) [reply .github/workflows/check_interface_compatibility.yml]: the latest version has enough comment to explain it

File: .github/workflows/check_interface_compatibility.yml (hunk 2 above)
Comment (aahung) [reply .github/workflows/check_interface_compatibility.yml]: not in the final version

File: .github/workflows/check_interface_compatibility.yml (hunk 3 above)
Comment (hoffa) [reply .github/workflows/check_interface_compatibility.yml:60]: Why not enforce?

Comment (aahung) [reply .github/workflows/check_interface_compatibility.yml:60]: As it could be intentional

Comment (aahung) [reply .github/workflows/check_interface_compatibility.yml:60]: And this homemade script could generate false positives

**Review by GavinZZ — APPROVED** [review]

File: .github/workflows/check_interface_compatibility.yml [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+      - name: Post GitHub PR comment
+        if: ${{ steps.detect.outputs.exit_code != '0' }}
+        uses: actions/github-script@v6
+        with:
+          github-token: ${{secrets.GITHUB_TOKEN}}
+          script: |
+            const fs = require('fs');
+            const output = fs.readFileSync('${{ runner.temp }}/output.md', 'utf8')
+            github.rest.issues.createComment({
```
Comment (acristin) [inline .github/workflows/check_interface_compatibility.yml:68]: Will this comment block merges until resolved, as a review comment would?

**Review by acristin — APPROVED** [review]

File: .github/workflows/check_interface_compatibility.yml (hunk 6 above)
Comment (aahung) [reply .github/workflows/check_interface_compatibility.yml:68]: good idea, let me check if there is an API for that

File: .github/workflows/check_interface_compatibility.yml [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+          make init
+          # Recover bin/public_interface.py
+          cp ${{ runner.temp }}/public_interface.py bin/public_interface.py
+          bin/public_interface.py extract > ${{ runner.temp }}/interfaces.new.json
+    
+      - name: Detect compatibility breaking changes
+        id: detect
+        run: |
+          exit_code=0
+          output=$(bin/public_interface.py check ${{ runner.temp }}/interfaces.original.json ${{ runner.temp }}/interfaces.new.json) || exit_code=$?
```
Comment (hoffa) [inline .github/workflows/check_interface_compatibility.yml:48]: AFAIK you don't need the initialization or `||` thing. Just `exit_code=$?` on separate line should be enough.

File: .github/workflows/check_interface_compatibility.yml [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+permissions:
+  pull-requests: write
+
+jobs:
+  check-interface-compatibility:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v3
```
Comment (hoffa) [inline .github/workflows/check_interface_compatibility.yml:14]: For symmetry with the other checkout, a `name` here describing what it checks out would be nice.

File: .github/workflows/check_interface_compatibility.yml [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+          cp bin/public_interface.py ${{ runner.temp }}/public_interface.py
+
+      - name: Checkout the PR
+        uses: actions/checkout@v3
+        with:
+          ref: "refs/pull/${{ github.event.number }}/merge"
+
+      - name: Extract PR public interfaces
+        run: |
+          make init
```
Comment (hoffa) [inline .github/workflows/check_interface_compatibility.yml:39]: Why is this needed again? Aren't all the dependencies from the base branch code, since that's the only script that's used?

File: .github/workflows/check_interface_compatibility.yml (hunk 6 above)
Comment (aahung) [reply .github/workflows/check_interface_compatibility.yml:68]: to create a PR comment that requires resolution, `path` and `line` are needed. Maybe it is too much work

https://octokit.github.io/rest.js/v19#pulls-create-review-comment

File: .github/workflows/check_interface_compatibility.yml [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+            github.rest.issues.createComment({
+              issue_number: context.issue.number,
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: output
+            })
+
+      - name: Determine the success/failure of the workflow
+        run: |
+          if [ ${{ steps.detect.outputs.exit_code }} -ne 0 ]; then
```
Comment (hoffa) [inline .github/workflows/check_interface_compatibility.yml:77]: Can this be like L61? So the UI shows nicer that the step is skipped.

Comment (hoffa) [reply .github/workflows/check_interface_compatibility.yml:77]: nit: and consider using a regular 'boolean' like in https://github.com/aws/serverless-application-model/blob/53865b59fbd9a3caae57159236378420a4420276/.github/workflows/schema.yml#L32 so it's arguably easier to read, as we really only care about whether there's a breaking change, not about exit codes.

File: .github/workflows/check_interface_compatibility.yml (hunk 9 above)
Comment (aahung) [reply .github/workflows/check_interface_compatibility.yml:39]: in case there is dependency change and module cannot be imported if samtranslator depends on a new pypi package.

File: .github/workflows/check_interface_compatibility.yml (hunk 8 above)
Comment (aahung) [reply .github/workflows/check_interface_compatibility.yml]: updated

File: .github/workflows/check_interface_compatibility.yml (hunk 7 above)
Comment (aahung) [reply .github/workflows/check_interface_compatibility.yml:50]: same issue as we saw in the release script, if `bin/public_interface.py` exit 1, the whole job stops without `||`

File: .github/workflows/check_interface_compatibility.yml (hunk 10 above)
Comment (aahung) [reply .github/workflows/check_interface_compatibility.yml]: good point, updated.

Comment (aahung) [reply .github/workflows/check_interface_compatibility.yml]: ok it worked https://github.com/aws/serverless-application-model/actions/runs/4130330209/jobs/7136902172

File: .github/workflows/check_interface_compatibility.yml (hunk 6 above)
Comment (acristin) [reply .github/workflows/check_interface_compatibility.yml:72]: Yeah, that's what I concluded as well.

Wow, that documentation seems much better than the official docs; happy to learn of it

File: .github/workflows/check_interface_compatibility.yml [hunk 11]
```diff
@@ -0,0 +1,83 @@
+name: Check interface compatibility
```
Comment (hoffa) [inline .github/workflows/check_interface_compatibility.yml:1]: Actually I think we need a way to run this easily locally, e.g. `make compat`. Currently it'll probably get annoying to not be able to reproduce and have big comments in PRs.

We could store the JSON in the repository and compare against that, no?

**Review by hoffa — APPROVED** [review]
